go/proof.go

```diff
@@ -309,6 +309,42 @@ func getPadding(spec *InnerSpec, branch int32) (minPrefix, maxPrefix, suffix int
 	return
 }
 
+// leftBranchesAreEmpty returns true if the padding bytes correspond to all empty children
+// on the left side of this branch, ie. it's a valid placeholder on a leftmost path
+func leftBranchesAreEmpty(spec *InnerSpec, op *InnerOp, branch int32) bool {
+	idx := getPosition(spec.ChildOrder, branch)
+	// compare the prefix bytes with the appropriate number of empty children
+	leftChildren := len(spec.ChildOrder) - 1 - idx
+	actualPrefix := len(op.Prefix) - leftChildren*int(spec.ChildSize)
+	if actualPrefix < 0 {
+		return false
+	}
+	for i := 0; i < leftChildren; i++ {
+		from := actualPrefix + i*int(spec.ChildSize)
+		if !bytes.Equal(spec.EmptyChild, op.Prefix[from:from+int(spec.ChildSize)]) {
+			return false
+		}
+	}
+	return true
+}
+
+// rightBranchesAreEmpty returns true if the padding bytes correspond to all empty children
+// on the right side of this branch, ie. it's a valid placeholder on a rightmost path
+func rightBranchesAreEmpty(spec *InnerSpec, op *InnerOp, branch int32) bool {
+	idx := getPosition(spec.ChildOrder, branch)
+	// compare the suffix bytes with the appropriate number of empty children
+	if len(op.Suffix) != idx*int(spec.ChildSize) {
```

Comment:
Again this is backwards.

`idx` is the number of left children, not the number of right children (items in suffix)

In this case, we want:

```go
idx := getPosition(spec.ChildOrder, branch)
rightBranches := len(spec.ChildOrder) - 1 - idx
if len(op.Suffix) != rightBranches*int(spec.ChildSize) {
  return false
}
for i := 0; i < rightBranches; i++ {
  // ...
}
```